stateproof/signer.go

```diff
@@ -0,0 +1,176 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package stateproof
+
+import (
+	"time"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto/merklesignature"
+	"github.com/algorand/go-algorand/data/account"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// sigFromAddr encapsulates a signature on a block header, which
+// will eventually be used to form a state proof for that
+// block.
+type sigFromAddr struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	SignerAddress basics.Address            `codec:"a"`
+	Round         basics.Round              `codec:"r"`
+	Sig           merklesignature.Signature `codec:"s"`
+}
+
+func (spw *Worker) signer(latest basics.Round) {
+	nextRnd := spw.nextStateProofRound(latest)
+	for { // Start signing StateProofs from nextRnd onwards
+		select {
+		case <-spw.ledger.Wait(nextRnd):
+			hdr, err := spw.ledger.BlockHdr(nextRnd)
+			if err != nil {
+				spw.log.Warnf("spw.signer(): BlockHdr(next %d): %v", nextRnd, err)
+				time.Sleep(1 * time.Second)
+				nextRnd = spw.nextStateProofRound(spw.ledger.Latest())
+				continue
+			}
+			spw.signBlock(hdr)
+			spw.signedBlock(nextRnd)
+			nextRnd++
+
+		case <-spw.ctx.Done():
+			spw.wg.Done()
+			return
+		}
+	}
+}
+
+func (spw *Worker) nextStateProofRound(latest basics.Round) basics.Round {
+	var nextrnd basics.Round
+
+	for {
+		latestHdr, err := spw.ledger.BlockHdr(latest)
+		if err != nil {
+			spw.log.Warnf("spw.signer(): BlockHdr(latest %d): %v", latest, err)
+			time.Sleep(1 * time.Second)
+			latest = spw.ledger.Latest()
+			continue
+		}
+
+		nextrnd = latestHdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound
+		if nextrnd == 0 {
+			// State proofs are not enabled yet.  Keep monitoring new blocks.
+			nextrnd = latest + 1
+		}
+		break
+	}
+
+	return nextrnd
+}
+
+func (spw *Worker) signBlock(hdr bookkeeping.BlockHeader) {
```

Comment:
signBlock is misleading imo, like we sign a block and embed a signature into it. In fact, we generate a signature for a header and store it into db. maybe call it `buildHdrSignature` or something?